### PR TITLE
Enable write-ahead logging and memory mapped IO

### PR DESCRIPTION
**Pull request by samuel-w**
Closes https://github.com/duplicati/duplicati/issues/3292

### CODE REVIEW

**Comment by samuel-w** [comment]
We could make it so that it only happens on first creation, but then it wouldn't apply to existing DBs.

File: Duplicati/Library/SQLiteHelper/SQLiteLoader.cs [hunk 1]
```diff
@@ -271,6 +271,13 @@ private static void OpenSQLiteFile(System.Data.IDbConnection con, string path)
             con.ConnectionString = "Data Source=" + path;
             con.Open();
 
+            // Enable write-ahead logging and memory mapped IO
+            using (System.Data.IDbCommand command = con.CreateCommand())
+            {
+                command.CommandText = "PRAGMA journal_mode = WAL;PRAGMA mmap_size=268435456;";
```
Comment (samuel-w) [inline Duplicati/Library/SQLiteHelper/SQLiteLoader.cs:277]: We could make this only run on first DB creation, but then it wouldn't work for existing DBs.

**Comment by warwickmm** [comment]
I have little knowledge in this area, but the [documentation](https://sqlite.org/wal.html) mentions

> WAL normally requires that the VFS support shared-memory primitives. (Exception: WAL without shared memory) The built-in unix and windows VFSes support this but third-party extension VFSes for custom operating systems might not.

Any idea how this might affect Synology devices, etc.?

File: Duplicati/Library/SQLiteHelper/SQLiteLoader.cs [hunk 2]
```diff
@@ -271,6 +271,13 @@ private static void OpenSQLiteFile(System.Data.IDbConnection con, string path)
             con.ConnectionString = "Data Source=" + path;
             con.Open();
 
+            // Enable write-ahead logging and memory mapped IO
+            using (System.Data.IDbCommand command = con.CreateCommand())
+            {
+                command.CommandText = "PRAGMA journal_mode = WAL;PRAGMA mmap_size = 8589934592;";
```
[the remark this answers is not in the document]
Comment (samuel-w) [reply Duplicati/Library/SQLiteHelper/SQLiteLoader.cs:277]: [The documentation](https://www.sqlite.org/mmap.html) says to enable memory mapped IO, the size should be some large value, so its some arbitrarily large value I came up with.

**Comment by mr-russ** [comment]
> Any idea how this might affect Synology devices, etc.?

SQLite will automatically fallback to journal meth5if filesystem does not support wal.  What filesystem does Synology usually use when it's installed there?

File: Duplicati/Library/SQLiteHelper/SQLiteLoader.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (mr-russ) [reply Duplicati/Library/SQLiteHelper/SQLiteLoader.cs:277]: This is the maximum virtual memory mapping you can have.  No real memory is used.  Duplicati still supports 32bit, so 2GB is probably the maximum to allow memory for the rist of the process. But my 32bit knowledge is fading fast.

**Comment by mr-russ** [comment]
@samuel-w Are you still wanting to pursue getting this merged, or prefer someone took over any changes?

**Pull request closed without merging**